### PR TITLE
Update Go to version 1.16.1 & rules_go to version 0.27

### PR DESCRIPTION
**Pull request by iholder-redhat**
**What this PR does / why we need it**:
The update has two main motivations:
1) Fix the crypto bug which is listed here: https://bugzilla.redhat.com/show_bug.cgi?id=1929069. As written in bug description, this bug is fixed in Go upstream: golang/go@d95ca91
    
2) To ensure compatibility and align with OCP as they are moving to OpenShift 4.9.

Bazel has also been updated to support newer Go version.

**Special notes for your reviewer**:
I merged a [PR to bump golang version to 1.14.14](https://github.com/kubevirt/kubevirt/pull/5102). Later, @rmohr figured that `cannot allocate memory` messages pop up from different places and suspected they are caused by the version bump. Therefore, I have merged a [PR to revert the golang vervion bump](https://github.com/kubevirt/kubevirt/pull/5287).

Then, with the help of @rmohr , I performed several experiments to reveal differences (in terms of resource usage) between clusters that run the different versions. Unfortunately, there was no difference at all between the clusters so I couldn't trace the root cause.

Since we now need to use version 1.16.1 (not 1.14.14) I will try once again to bump the version. If the error will pop up again I will continue to try and understand its root cause.

**Release note**:

```release-note
Go version updated to version 1.16.1.
Bezel version updated to version 0.27.
```

### CODE REVIEW

File: staging/src/kubevirt.io/client-go/api/v1/schema.go [hunk 1]
```diff
@@ -189,7 +189,7 @@ type DomainSpec struct {
 	Memory *Memory `json:"memory,omitempty"`
 	// Machine type.
 	// +optional
-	Machine Machine `json:"machine,omitempty"`
+	Machine *Machine `json:"machine,omitempty"`
```
Comment (xpivarc) [inline staging/src/kubevirt.io/client-go/api/v1/schema.go:192]: What is the reason for this change?

Comment (rmohr) [reply staging/src/kubevirt.io/client-go/api/v1/schema.go:192]: The reason are changes in the validation. I asked @iholder-redhat to make it a pointer (since it is optional anyways), because an empty `type` on the machine was not accepted. But I checked now our example VMIs:

```
---
apiVersion: kubevirt.io/v1
kind: VirtualMachineInstance
metadata:
  labels:
    special: vmi-nocloud
  name: vmi-nocloud
spec:
  domain:
    devices:
      disks:
      - disk:
          bus: virtio
        name: containerdisk
      - disk:
          bus: virtio
        name: cloudinitdisk
      - disk:
          bus: virtio
        name: emptydisk
    machine:
      type: ""
    resources:
      requests:
        memory: 128Mi
  terminationGracePeriodSeconds: 0
  volumes:
  - containerDisk:
      image: registry:5000/kubevirt/cirros-container-disk-demo:devel
    name: containerdisk
  - cloudInitNoCloud:
      userData: |
        #!/bin/sh

        echo 'printed from cloud-init userdata'
    name: cloudinitdisk
  - emptyDisk:
      capacity: 2Gi
    name: emptydisk
```

We have 

```
    machine:
      type: ""
```

in there and we accepted it. That means that we need  continue accepting it. @iholder-redhat I think that means reverting this part and adding the optional the type in Machine.

**Comment by rmohr** [comment]
here on what the e2e test are failing: 

```
15:48:17:   admission webhook "kubevirt-update-validator.kubevirt.io" denied the request: status.generations.hash in body must be of type string: "null", status.generations.namespace in body must be of type string: "null"
```

**Comment by iholder-redhat** [comment]
> here on what the e2e test are failing:
> 
> ```
> 15:48:17:   admission webhook "kubevirt-update-validator.kubevirt.io" denied the request: status.generations.hash in body must be of type string: "null", status.generations.namespace in body must be of type string: "null"
> ```

Yes I saw that, thanks!

The root cause is that `GenerationStatus` struct, which is defined in OpenShift, has a `hash` field which is said to be optional in the comments but isn't actually optional. Since there was a bug in Go regarding enforcing mandatory fields everything was fine until now - but now validation fails (as it should be) because hash is empty.

I'm willing to make a PR for openshift to fix this. The question is whether we want to wait for this PR to be merged and than update our library or in the mean time I can replicate the struct from OpenShift to Kubevirt. I'll now try doing this since it's probably not too hard.

**Comment by xpivarc** [comment]
@iholder-redhat I think you can work around the problem with openshift struct in `LoadOpenAPISpec` (At least for now).

File: staging/src/kubevirt.io/client-go/api/v1/schema.go (hunk 1 above)
Comment (xpivarc) [reply staging/src/kubevirt.io/client-go/api/v1/schema.go:192]: Thanks for the explanation and yes we need to add it there.

**Comment by rmohr** [comment]
> @iholder-redhat I think you can work around the problem with openshift struct in `LoadOpenAPISpec` (At least for now).

I am still a little bit confused, shouldn't we have consolidated the "optional" check to the `+optional` annotation in the openapi spec, which should make `omitempty` mostly a nice to have?

**Comment by iholder-redhat** [comment]
> > @iholder-redhat I think you can work around the problem with openshift struct in `LoadOpenAPISpec` (At least for now).
> 
> I am still a little bit confused, shouldn't we have consolidated the "optional" check to the `+optional` annotation in the openapi spec, which should make `omitempty` mostly a nice to have?

You can look here: https://github.com/openshift/api/blob/master/operator/v1/types.go#L139.
In OpenShift the comment says `hash is an optional field ...` but there is no actual `+optional` comment there. Clearly this field is meant to be optional but de-facto it's mandatory

File: staging/src/kubevirt.io/client-go/api/v1/schema.go (hunk 1 above)
Comment (iholder-redhat) [reply staging/src/kubevirt.io/client-go/api/v1/schema.go:192]: Hey @rmohr,

We did have
```
    machine:
      type: ""
````
But it actually was an unintended mistake as `type` is supposed to be mandatory according to our schema definitions. Seems like because of Go's bug (not enforcing mandatory fields properly) we didn't notice that our example VMIs are actually invalid according to our own definitions.

I guess we'll just adapt this mistake to not break backwards comparability? Can this mistake, i.e. machine without a defined type, cause problems in the future?

**Comment by rmohr** [comment]
> You can look here: https://github.com/openshift/api/blob/master/operator/v1/types.go#L139.
> In OpenShift the comment says `hash is an optional field ...` but there is no actual `+optional` comment there. Clearly this field is meant to be optional but de-facto it's mandatory

Ah if `+optional` is missing there then the model in my mind matches reality again :)

**Comment by iholder-redhat** [comment]
/cc @ashleyschuett 

@rmohr @xpivarc Another issue popped up.
`hash` is not our only problem - turns out that `Namespace` field is also sometimes empty although it is mandatory. The reason for this is we track CRDs which do not have a namespace.

I can just make `Namespace` optional but then we'll diverge from OpenShift's definitions. Another naive option is define a dummy namespace for CRDs, what garbage values are not ideal as well. What do you think is the right approach here?

File: staging/src/kubevirt.io/client-go/api/v1/schema.go (hunk 1 above)
Comment (rmohr) [reply staging/src/kubevirt.io/client-go/api/v1/schema.go:192]: > But it actually was an unintended mistake as `type` is supposed to be mandatory according to our schema definitions. Seems like because of Go's bug (not enforcing mandatory fields properly) we didn't notice that our example VMIs are actually invalid according to our own definitions.

I don't see "machine"  required in any spec on master. Where do you see that?

Comment (rmohr) [reply staging/src/kubevirt.io/client-go/api/v1/schema.go:192]: > I guess we'll just adapt this mistake to not break backward comparability? Can this mistake, i.e. machine without a defined type, cause problems in the future?

You can make it optional. It is not beautiful, but that is fine since we anyway default it always. After the VMI went through the webhooks, there is no case where no machine type is set.

**Comment by rmohr** [comment]
> I can just make `Namespace` optional but then we'll diverge from OpenShift's definitions.

From which definition?

**Comment by iholder-redhat** [comment]
> > I can just make `Namespace` optional but then we'll diverge from OpenShift's definitions.
> 
> From which definition?

This one: https://github.com/openshift/api/blob/master/operator/v1/types.go#L128

**Comment by rmohr** [comment]
> This one: https://github.com/openshift/api/blob/master/operator/v1/types.go#L128

we could override the openapi specification on a few places in our project.


Here how `Quantity` is defining its openapi structure:

```
// OpenAPISchemaType is used by the kube-openapi generator when constructing
// the OpenAPI spec of this type.
//
// See: https://github.com/kubernetes/kube-openapi/tree/master/pkg/generators
func (_ Quantity) OpenAPISchemaType() []string { return []string{"string"} }

// OpenAPISchemaFormat is used by the kube-openapi generator when constructing
// the OpenAPI spec of this type.
func (_ Quantity) OpenAPISchemaFormat() string { return "" }
```

File: pkg/virt-operator/resource/apply/generations.go [hunk 2]
```diff
@@ -3,12 +3,12 @@ package apply
 import (
 	"fmt"
 
+	v12 "kubevirt.io/client-go/api/v1"
```
Comment (dhiller) [inline pkg/virt-operator/resource/apply/generations.go:6]: Maybe better `kubevirtclientv1`?

**Review by dhiller — COMMENTED** [review]
I understand this is still a work in progress, so I haven't had a look at the draft commit. The other commits look good to me, one nit inline.

File: manifests/generated/kv-resource.yaml [hunk 3]
```diff
@@ -1112,6 +1112,11 @@ spec:
                     resource:
                       description: resource is the resource type of the thing you're tracking
                       type: string
+                  required:
```
Comment (xpivarc) [inline manifests/generated/kv-resource.yaml:1115]: I am wondering why this appears here.

File: pkg/virt-operator/resource/apply/generations.go (hunk 2 above)
Comment (rmohr) [reply pkg/virt-operator/resource/apply/generations.go:6]: another nice option is `k6tv1`.

File: staging/src/kubevirt.io/client-go/api/v1/schema.go [hunk 4]
```diff
@@ -189,7 +189,7 @@ type DomainSpec struct {
 	Memory *Memory `json:"memory,omitempty"`
 	// Machine type.
 	// +optional
-	Machine Machine `json:"machine,omitempty"`
+	Machine *Machine `json:"machine"`
```
Comment (xpivarc) [inline staging/src/kubevirt.io/client-go/api/v1/schema.go:192]: You need to keep `omitempty` here otherwise nil `Machine` will result in `null`  in json. This will be denied by webhook.

**Review by xpivarc — CHANGES_REQUESTED** [review]

File: tests/vmi_configuration_test.go [hunk 5]
```diff
@@ -1652,17 +1652,28 @@ var _ = Describe("[sig-compute]Configurations", func() {
...
-			vmi.Spec.Domain.Machine.Type = ""
+			vmi.Spec.Domain.Machine = nil
+			tests.RunVMIAndExpectLaunch(vmi, 30)
+			runningVMISpec, err := tests.GetRunningVMIDomainSpec(vmi)
+
+			Expect(err).ToNot(HaveOccurred())
+			Expect(runningVMISpec.OS.Type.Machine).To(ContainSubstring("q35"))
+		})
+
+		It("should allow creating VM defined with Machine with an empty Type", func() {
```
Comment (iholder-redhat) [inline tests/vmi_configuration_test.go:1673]: What would you suggest to do here regarding test ids?

File: staging/src/kubevirt.io/client-go/api/v1/schema_test.go [hunk 6]
```diff
@@ -67,9 +67,6 @@ var exampleJSONFmt = `{
         ],
         "dedicatedCpuPlacement": true
       },
-      "machine": {
```
[the remark this answers is not in the document]
Comment (dhiller) [reply staging/src/kubevirt.io/client-go/api/v1/schema_test.go:70]: Think so too. Maybe we can have two tests, one with the original and another one with the missing section?

[the remark this answers is not in the document]
Comment (iholder-redhat) [reply staging/src/kubevirt.io/client-go/api/v1/schema_test.go:70]: Why do think it needs to stay? Our example VMIs are not defined with a `machine` while it's now marked with `omitempty`

p.s. I've added a test to ensure that the old JSON format is also accepted as valid

**Comment by iholder-redhat** [comment]
@rmohr @dhiller 
looks like we have a problem..
[test is failing](https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/5495/pull-kubevirt-e2e-k8s-1.18/1389240476791476224) on `Unexpected Warning event received: testvmi-jvdvd,a3b23700-6651-41fe-b78b-3e0a4523bb87: failed to get image info: failed to invoke qemu-img: exit status 2: 'fatal error: runtime: cannot allocate memory`

**Comment by iholder-redhat** [comment]
> @rmohr @dhiller
> looks like we have a problem..
> [test is failing](https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/5495/pull-kubevirt-e2e-k8s-1.18/1389240476791476224) on `Unexpected Warning event received: testvmi-jvdvd,a3b23700-6651-41fe-b78b-3e0a4523bb87: failed to get image info: failed to invoke qemu-img: exit status 2: 'fatal error: runtime: cannot allocate memory`

Added another 100Mi to virt-chroot execution as @rmohr suggested. Let's hope.

**Comment by rmohr** [comment]
This is weird. I checked https://search.ci.kubevirt.io/?search=%3Cv1.MigrationMode%3E%3A+PostCopy&maxAge=336h&context=1&type=bug%2Bjunit&name=&excludeName=&maxMatches=5&maxBytes=20971520&groupBy=job

 and I see that the error seen here is happening since at least 14 days (max search history), and there I saw it on a periodic, so not related to this PR. But in your PR it happens a lot more. I think we have to investigate that.

This is the error:

```
Expected
    <v1.MigrationMode>: PreCopy
to equal
    <v1.MigrationMode>: PostCopy
```

File: staging/src/kubevirt.io/client-go/api/v1/types.go [hunk 7]
```diff
@@ -1534,7 +1532,7 @@ type KubeVirtStatus struct {
 	ObservedDeploymentID                    string              `json:"observedDeploymentID,omitempty" optional:"true"`
 	OutdatedVirtualMachineInstanceWorkloads *int                `json:"outdatedVirtualMachineInstanceWorkloads,omitempty" optional:"true"`
 	// +listType=atomic
-	Generations []operatorsv1.GenerationStatus `json:"generations,omitempty" optional:"true"`
+	Generations []GenerationStatus `json:"generations,omitempty" optional:"true"`
```
Comment (davidvossel) [inline staging/src/kubevirt.io/client-go/api/v1/types.go:1535]: fyi, We're considering go with a changesimilar to this earlier due to an relase-0.41 blocker, see https://github.com/kubevirt/kubevirt/pull/5609

**Comment by rmohr** [comment]
@vladikr 

regarding this migration error which also happens sporadically outside of this PR. Can it be that sometimes the migration finishes so fast, that `preCopy` mode is never reported from the source virt-handler because either libvirt does not have the chance to change the mode fast enough, or that the source virt-handler does not receive the notification about the post-to-pre copy switch in time?

**Comment by vladikr** [comment]
> @vladikr
> 
> regarding this migration error which also happens sporadically outside of this PR. Can it be that sometimes the migration finishes so fast, that `preCopy` mode is never reported from the source virt-handler because either libvirt does not have the chance to change the mode fast enough, or that the source virt-handler does not receive the notification about the post-to-pre copy switch in time?

hmm, I see that `[test_id:5004] with postcopy` doesn't even run `stress`... the migration finishes in much less than a second (1 second at least is required for us to switch to post copy)

```
"DomainLifecycle event 5 with reason 3 received","pos":"client.go:401","timestamp":"2021-05-05T18:30:41.850294Z"
"   kubevirt domain status: Shutoff(5):Migrated(4)","pos":"client.go:262","timestamp":"2021-05-05T18:30:41.855145Z"
```

` [test_id:4747] ` runs stress, but I'm not sure if it actually runs... perhaps we are asking for too much memory for stress to allocate. Not sure what has changed with this PR, but maybe we should try allocation 600mb instead of 800mb

File: staging/src/kubevirt.io/client-go/api/v1/types.go (hunk 7 above)
Comment (iholder-redhat) [reply staging/src/kubevirt.io/client-go/api/v1/types.go:1535]: Thanks

**Comment by iholder-redhat** [comment]
@rmohr @vladikr please be aware that test `5004` is getting into quarantine for being too flakie. I've asked in the PR if tests `5004` and `4747` might be related and currently we're not sure about that.

**Comment by vladikr** [comment]
> @rmohr @vladikr please be aware that test `5004` is getting into quarantine for being too flakie. I've asked in [the PR](https://github.com/kubevirt/kubevirt/pull/5658) if tests `5004` and `4747` might be related and currently we're not sure about that.

trying to test this locally now.

**Comment by vladikr** [comment]
ohh I know what the problem is.
Somehow the current version on libvirt-go doesn't have a fix I've introduced in [1] and [2] had the updated version of libvirt with this change 
[1] https://gitlab.com/libvirt/libvirt-go/-/merge_requests/27
[2] https://github.com/kubevirt/kubevirt/pull/5590/files#diff-221c7fcbf31c284c7bc2ba0a899459cd0b75252017012f9b82d33ef8e3e96f52R3350

But not I don't see in on this branch now....

@rmohr @andreabolognani @fgimenez 
The tests that were verifying the fix are still in quarantine.

This also will affect https://github.com/kubevirt/kubevirt/pull/5658
I think we just need to fix the issue and not to put more migration tests to quarantine.

**Comment by vladikr** [comment]
ah @iholder-redhat  you've reverted it some how 
with 
```
commit 025e51a41e261d68fb2e4c8dc3a00124b2bb053b
Author: Itamar Holder <iholder@redhat.com>
Date:   Wed May 12 14:29:29 2021 +0300

    Update go-openapi version
    
    This update is essential to align with open-api move to
    Go 1.16. Go had several bugs that were fixed which cased
    new bugs in open-api that were fixed in later versions.
    
    Signed-off-by: Itamar Holder <iholder@redhat.com>
```

**Comment by vladikr** [comment]
looks like it's all my fault.
We need https://github.com/kubevirt/kubevirt/pull/5601 to get merged before this one, otherwise make dep-update will override the libvirt-go change.

**Comment by iholder-redhat** [comment]
> otherwise make dep-update will override the libvirt-go change.

Yes, I think this is what's happened. I rebased (then had your changes), then to make sure I didn't forget something I've done `make generate && make deps-update` and accidentally override your changes. Thanks so much for your help!!

BTW, since https://github.com/kubevirt/kubevirt/pull/5658 is going to get merged very soon, I suggest to `/hold` it if we don't want it in.

**Comment by rmohr** [comment]
> BTW, since #5658 is going to get merged very soon, I suggest to `/hold` it if we don't want it in.

The tests are failing on flakefinder. flakefinder only shows results of merged code, which means that it is also flaky on master where still the fix is present.

**Comment by iholder-redhat** [comment]
@vladikr test 4747 keeps failing on same thing as before:
```
Expected
    <v1.MigrationMode>: PreCopy
to equal
    <v1.MigrationMode>: PostCopy
```

What do you suggest? Should I retest again?

**Comment by rmohr** [comment]
> What do you suggest? Should I retest again?

I would suggest to rebase locally and run the tests locally to see if things are better now.

**Comment by vladikr** [comment]
> @vladikr test 4747 keeps failing on same thing as before:

Just wondering if you have checked that the GetJobStats has the jobType field now, as below?

```
      params := DomainJobInfo{
              Type: DomainJobType(jobtype),
      }
```

And yes, we should test it locally...

**Comment by vladikr** [comment]
> @vladikr test 4747 keeps failing on same thing as before:

I see that vendor/libvirt.org/libvirt-go/domain.go on this PR is not updated. perhaps you should run make dep-update again?

**Comment by iholder-redhat** [comment]
> > @vladikr test 4747 keeps failing on same thing as before:
> 
> I see that vendor/libvirt.org/libvirt-go/domain.go on this PR is not updated. perhaps you should run make dep-update again?

Ah, you're right.
Wasn't that supposed to happen in the [PR that bumped libvirt](https://github.com/kubevirt/kubevirt/pull/5601)? Anyways I've ran `make deps-update` and added the changes to PR.

@rmohr Ran the test locally for 5 times - passed 100% of the times. Let's hope it's the last re-push for this PR :)
